pkg/platform/k8s/pod_controller.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// PodReconciler contain information on the clusters pods.
+type PodReconciler struct {
+	client.Client
+	ipToPod map[string]metav1.ObjectMeta
+	logger  *logrus.Entry
+}
+
+// Reconcile watches all pods events and updates the PodReconciler.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var pod corev1.Pod
+	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			// the pod was deleted.
+			r.deletePod(req.NamespacedName.Name, req.NamespacedName.Namespace)
+			return ctrl.Result{}, nil
+		}
+		r.logger.Error(err, "unable to fetch Pod")
+		return ctrl.Result{}, err
+	}
+
+	r.updatePod(pod)
+	return ctrl.Result{}, nil
+}
+
+// deletePod deletes pod to ipToPod list.
+func (r *PodReconciler) deletePod(name, namespace string) {
+	for key, pod := range r.ipToPod {
+		if pod.Name == name && pod.Namespace == namespace {
+			delete(r.ipToPod, key)
+		}
+	}
+}
+
+// updatePod adds or updates pod to ipToPod list.
+func (r *PodReconciler) updatePod(pod corev1.Pod) {
+	for _, ip := range pod.Status.PodIPs {
+		// ignoring host-networked Pod IPs
+		if ip.IP != pod.Status.HostIP {
+			r.ipToPod[ip.IP] = pod.ObjectMeta
+		}
+	}
+}
+
+// getLabelFromIP return all the labels for specific ip.
+func (r *PodReconciler) getLabelFromIP(ip string) map[string]string {
```

Comment:
See earlier comment and retaining and returning more than just labels (e.g., name and namespace can also be used in the policies)

Reply:
I do not think we would get the IP address as a string from the data plane. It is more likely a `[]byte` which needs to be parsed. While you could delegate it to the caller, it might be better done here to keep a uniform interface to the Go and Envoy implementations.
You should, at least, document the string format (e.g., IPv4, v6, ...)

Reply:
I think it is the responsibility of the controlplane to convert the bytes data to the correct IP format.